src/core/sock/sockinfo_tcp.cpp

```diff
@@ -6046,3 +6055,178 @@ inline bool sockinfo_tcp::handle_bind_no_port(int &bind_ret, in_port_t in_port,
 
     return CONTINUE_WITH_BIND;
 }
+
+int sockinfo_tcp::tcp_tx_express(const struct iovec *iov, unsigned iov_len, uint32_t mkey,
+                                 xlio_express_flags flags, void *opaque_op)
+{
+    if (unlikely(!is_connected_and_ready_to_send())) {
+        return -1;
+    }
+
+    pbuf_desc mdesc;
+
+    switch (flags & XLIO_EXPRESS_OP_TYPE_MASK) {
+    case XLIO_EXPRESS_OP_TYPE_DESC:
+        mdesc.attr = PBUF_DESC_EXPRESS;
+        break;
+    case XLIO_EXPRESS_OP_TYPE_FILE_ZEROCOPY:
+        mdesc.attr = PBUF_DESC_MDESC;
+        break;
+    default:
+        return -1;
+    };
+    mdesc.mkey = mkey;
+    mdesc.opaque = opaque_op;
+
+    int bytes_written = 0;
+
+    lock_tcp_con();
+
+    err_t err;
+    for (unsigned i = 0; i < iov_len; ++i) {
+        err = tcp_write_express(&m_pcb, iov[i].iov_base, iov[i].iov_len, &mdesc);
+        if (err != ERR_OK) {
+            /* The only error in tcp_write_express is a memory error
+             * In this version we don't implement any error recovery or avoidance
+             * mechanism and an error at this stage is irrecoverable.
+             * The considered alternatives are:
+             *  - Setting the socket an error state (this is the one we chose here)
+             *  - Rolling back any written buffers, i.e. recovering
```

Comment:
It seems that tcp_write_express already reverts back the segments on memory issue. So i guess we could just return error.